source/nsdynmemLIB/nsdynmemLIB.c

```diff
@@ -155,49 +156,58 @@ static int heap_alloc_internal_check(int16_t alloc_size)
 static int8_t ns_sector_validate(int *sector_start, int direction)
 {
     int8_t ret_val = -1;
-    int size_start, size_end;
-    if (direction) {
-        //Up
-        size_start = *sector_start++;
-        if (size_start < 0) {
-            sector_start += -(size_start);
-        } else {
-            sector_start += size_start;
-        }
-        size_end = *sector_start;
-    } else {
-        size_end = *sector_start--;
-        if (size_end < 0) {
-            sector_start -= -(size_end);
-        } else {
-            sector_start -= size_end;
-        }
-        size_start = *sector_start;
-    }
-    if (size_start == size_end) {
+    int size_start;
+//    int size_end;
+    int multiplier = direction?1:-1;
+    size_start = *sector_start;
+    sector_start += multiplier;
+    multiplier = size_start<0?-multiplier:multiplier;
+    sector_start += multiplier*size_start;
+
+//    if (direction) {
+//        //Up
+//        size_start = *sector_start++;
+//        if (size_start < 0) {
+//            sector_start += -(size_start);
+//        } else {
+//            sector_start += size_start;
+//        }
+//        size_end = *sector_start;
+//    } else {
+//        size_end = *sector_start--;
+//        if (size_end < 0) {
+//            sector_start -= -(size_end);
+//        } else {
+//            sector_start -= size_end;
+//        }
+//        size_start = *sector_start;
+//    }
+//    if (size_start == size_end) {
+    if (size_start == *sector_start) {
         ret_val = 0;
     }
     return ret_val;
 }
 #endif
 
-void *ns_dyn_mem_alloc(int16_t alloc_size)
+static void *ns_dyn_mem_internal_alloc(int16_t alloc_size, bool dir_up)
 {
-#ifndef STANDARD_MALLOC
-
-    int *ptr = heap_main_end;
+    int *ptr = dir_up?heap_main:heap_main_end;
```

Comment:
Like the concept, but suggest passing +1/-1 in the first place, if you're going to do the multiplication thing.